### ISSUE TITLE
CSP2: Default value of default-src is incorrect

### ISSUE THREAD

**Issue by marumari** [w3c/webappsec#514]
Here is what the current specification says:

> Let the default sources be the result of parsing the **_default-src_** directive’s value as a source list if a `default-src` directive is explicitly specified, and otherwise the U+002A ASTERISK character (*).

Which is incorrect, as it reads that these two statements are equivalent:

`Content-Security-Policy: default-src *; upgrade-insecure-requests`
`Content-Security-Policy: upgrade-insecure-requests`

The statement should probably read something like:

> Let the default sources be the result of parsing the **_default-src_** directive’s value as a source list if a `default-src` directive is explicitly specified, and otherwise the list of all possible sources.

(cc: @hillbrad, as @mikewest told me to)

**Comment by marumari** [comment]
Note that this appears to be fixed in CSP3, as near as I can tell.

**Comment by mikewest** [comment]
Yup. This is broken in CSP2. Although I think that we decided in some other bug that it didn't matter because CSP2 was so loosely specified that the `*` was never "enforced". Let me see if I can track that down.

**Comment by mikewest** [comment]
https://github.com/w3c/webappsec-csp/issues/9#issuecomment-146223446 is what I was thinking of. @hillbrad, @wseltzer, @dveditz: Is it worth revising CSP2 at this point? WDYT?

**Comment by marumari** [comment]
Well, I'm fine with this being loosely specified as well, just as long as it's clear that the two hand-wavingly vague definitions are distinct.

The browsers treat them very differently, and so this line was very confusing to me both when it came to writing documentation about CSP as well as when writing a CSP analyzer.

**Comment by mikewest** [comment]
> The browsers treat them very differently

Do you mean that browsers treat a missing `default-src` (like `Content-Security-Policy: script-src 'self'`) differently? Or that `*` means something very different from "nothing"?

**Comment by marumari** [comment]
The latter.  Not having a default-src is very different than default-src *, but the CSP2 spec implies that they are the same.

**Comment by hillbrad** [comment]
So, by "fixed in CSP3" does that mean we expect different behavior from the
same policy strings?  Or do you mean the existing behavior is more clearly
explained?

I think we can maybe punt on updating CSP2 if we verify by the test suite
that our implementations behave similarly.  If they have diverged due to
unclear language we need to fix it.

**Comment by marumari** [comment]
The behavior hasn't changed in CSP3, it's just that the CSP3 accurately describes the implemented behavior where as CSP2 does not.

**Comment by hillbrad** [comment]
Test shows that Chrome, Firefox and Safari Tech Preview are all compliant, so we can try to make this correction as a non-conformance affecting edit. (I'll manage that, @mikewest )

**Comment by mikewest** [comment]
You are the best, @hillbrad. Thank you!

**Comment by mikewest** [comment]
Fixed this in the editor's draft, using @marumari's suggestion. Brad, perhaps you could sneakily apply https://github.com/w3c/webappsec-csp/commit/bc9577dffd1fff0bcea3f23423d78a6f060b3c47 to the CR?
